packages/block-library/src/block/edit.js

```diff
@@ -21,9 +26,52 @@ import {
 	InspectorControls,
 	useBlockProps,
 	Warning,
+	privateApis as blockEditorPrivateApis,
 } from '@wordpress/block-editor';
+import { useState, useMemo } from '@wordpress/element';
+
+/**
+ * Internal dependencies
+ */
+import { unlock } from '../lock-unlock';
+
+const fullAlignments = [ 'full', 'wide', 'left', 'right' ];
+
+const useInferredLayout = ( blocks, parentLayout ) => {
+	const [ inferredAlignment, setInferredAlignment ] = useState();
 
-export default function ReusableBlockEdit( { attributes: { ref } } ) {
+	return useMemo( () => {
+		// Exit early if the pattern's blocks haven't loaded yet.
+		if ( ! blocks?.length ) {
+			return {};
+		}
+
+		let alignment = inferredAlignment;
+
+		// Only track the initial alignment so that temporarily removed
+		// alignments can be reapplied.
+		if ( inferredAlignment === undefined ) {
+			const isConstrained = parentLayout?.type === 'constrained';
+			const hasFullAlignment = blocks.some( ( block ) =>
+				fullAlignments.includes( block.attributes.align )
+			);
+
+			alignment = isConstrained && hasFullAlignment ? 'full' : null;
+			setInferredAlignment( alignment );
+		}
```

Comment:
The reason for using state here isn't super clear, IMO. Should this be wrapped in effect instead of a memo?

The `hasFullAlignment` can be derived outside the hook and used as a dependency instead of `blocks`. This will re-calculate memoized value less often.

Reply:
Thanks @Mamaduka, I've switched this to use a ref instead of state as @kevin940726 suggested in https://github.com/WordPress/gutenberg/pull/54416/commits/b8bb61949d9220d4de74760b5bb218c9f6a552dd.